src/app/contexts/Ledger/use-ledger.ts

```diff
@@ -0,0 +1,182 @@
+						return;
+					}
+					dispatch({ type: "remove" });
+				},
+				error: (e) => dispatch({ type: "failed", message: e.message }),
+				complete: () => void 0,
+			}),
+		[transport],
+	);
+
+	const importLedgerWallets = useCallback(
+		async (wallets: LedgerData[], coin: Coins.Coin, profile: Profile) => {
+			for (const { address, index } of wallets) {
+				const wallet = await profile
+					.wallets()
+					.importByAddress(address, coin.network().coin(), coin.network().id());
+				wallet.data().set(WalletFlag.LedgerIndex, index);
+			}
+			await persist();
+		},
+		[persist],
+	);
+
+	const connect = useCallback(
+		async (
+			coin: string,
+			network: string,
+			retryOptions: retry.Options = { retries: 50, randomize: false, factor: 1 },
+		) => {
+			dispatch({ type: "waiting" });
+			abortRetryRef.current = false;
+
+			const instance = await env.coin(coin, network);
+
+			try {
+				const slip44 = instance.config().get<number>("network.crypto.slip44");
+
+				const connectFn: retry.RetryFunction<void> = async (bail, attempts) => {
+					if (abortRetryRef.current && attempts > 1) {
+						bail(new Error("User aborted"));
+					}
+
+					await instance.ledger().connect(transport);
+					// Ensure that the app is accessible
+					await instance.ledger().getPublicKey(formatDerivationPath(slip44, 0));
+				};
+
+				await retry(connectFn, retryOptions);
+				dispatch({ type: "connected" });
+			} catch (connectError) {
+				instance.ledger().disconnect();
+				dispatch({ type: "failed", message: connectError.message });
+				throw connectError;
+			}
+		},
+		[transport, env],
+	);
+
+	const disconnect = useCallback(async (coin: Coins.Coin) => {
+		await coin.ledger().disconnect();
+		dispatch({ type: "disconnected" });
+	}, []);
+
+	const scanWallets = useCallback(
+		async (coin: string, network: string, profile: Profile, onChange?: (wallet: LedgerData) => void) => {
+			const wallets: LedgerData[] = [];
+
+			try {
+				await connect(coin, network);
+
+				const instance = await env.coin(coin, network);
+				const slip44 = instance.config().get<number>("network.crypto.slip44");
+
+				dispatch({ type: "busy" });
+
+				let hasMore = true;
+				let cursor = 0;
+
+				while (hasMore) {
+					const path = formatDerivationPath(slip44, cursor);
+					const publicKey = await instance.ledger().getPublicKey(path);
+					const address = await instance.identity().address().fromPublicKey(publicKey);
+
+					// Already imported
+					if (profile.wallets().findByPublicKey(publicKey)) {
+						cursor++;
+						continue;
+					}
+
+					let wallet: LedgerData;
+					try {
+						const identity = await instance.client().wallet(address);
+						wallet = {
+							address: identity.address(),
+							balance: identity.balance(),
+							index: cursor,
+						};
+					} catch {
+						// New Cold Wallet
+						wallet = {
+							address,
+							balance: BigNumber.ZERO,
+							index: cursor,
+							isNew: true,
+						};
+						hasMore = false;
```

Comment:
Doesn't this completely abort importing of any wallets if a single wallet can't be retrieved via API? This doesn't necessarily mean that the wallet doesn't exist, could be just a timeout or rate limiting.

Reply:
Should I check that the error is not 404 and retry?

Reply:
I would simply continue and store the wallets that failed in an array. Then when all wallets are imported we can tell the user that wallets X, Y and Z failed to be imported and let them retry because something like a rate limit timeout can persist for up to 60 seconds. Probably best to ask Oleg for a design.